File: 2024/day06Optimized/day06.js
// Optimized JavaScript implementation
'use strict';

// Performance measurement
console.time("time");

const fs = require("fs");
const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');

// Constants for better performance and readability
const DIR_UP = "^";
const DIR_DOWN = "v";
const DIR_LEFT = "<";
const DIR_RIGHT = ">";
const WALL = "#";
const OPEN = ".";

// Pre-computed movement and turn maps (avoids lookups in hot loops)
const nextPosMods = {
  [DIR_UP]: [-1, 0],
  [DIR_DOWN]: [1, 0],
  [DIR_LEFT]: [0, -1],
  [DIR_RIGHT]: [0, 1],
};

const turns = {
  [DIR_UP]: DIR_RIGHT,
  [DIR_DOWN]: DIR_LEFT,
  [DIR_LEFT]: DIR_UP,
  [DIR_RIGHT]: DIR_DOWN,
};

const filepath = "../inputs/day06.txt";

// Worker thread code
if (!isMainThread) {
  const { grid: workerGrid, guardPos, trodden } = workerData;
  
  let localCount = 0;
  for (const posStr of trodden) {
    const pos = posStr.split(':').map(Number);
    
    // Skip guard starting position
    if (pos[0] === guardPos[0] && pos[1] === guardPos[1]) {
      continue;
    }
    
    // Clone grid for this iteration (more efficient than modifying and restoring)
    const grid = workerGrid.map(row => [...row]);
    
    // Modify grid and check solution
    grid[pos[0]][pos[1]] = WALL;
    const completed = solve(grid);
    if (!completed) {
      localCount++;
    }
  }
  
  // Send result back to main thread
  parentPort.postMessage(localCount);
}
// Main thread code
else {
  main().catch(console.error);
}

async function main() {
  // Part 1
  const path = runPt1();
  const keys = Object.keys(path);
  console.log(`Part 1: ${keys.length}`);
  
  // Part 2
  await runPt2(keys);
  
  console.timeEnd("time");
}

function runPt1() {
  const grid = getGrid();
  return solve(grid);
}

async function runPt2(trodden) {
  const grid = getGrid();
  const guard = getGuard(grid);
  
  // Calculate thread pool size (75% of available CPU cores)
  const numCPUs = require('os').cpus().length;
  const numWorkers = Math.max(1, Math.floor(numCPUs * 0.75));
  
  // Split work into chunks for workers
  const chunkSize = Math.ceil(trodden.length / numWorkers);
  const chunks = [];
  
  for (let i = 0; i < numWorkers; i++) {
    const start = i * chunkSize;
    const end = Math.min(start + chunkSize, trodden.length);
    chunks.push(trodden.slice(start, end));
  }
  
  // Process chunks in parallel using worker threads
  const results = await Promise.all(chunks.map(chunk => 
    runWorker(grid, guard.pos, chunk)
  ));
  
  // Sum up results from all workers
  const count = results.reduce((sum, val) => sum + val, 0);
  console.log(`Part 2: ${count}`);
  return count;
}

function runWorker(grid, guardPos, trodden) {
  return new Promise((resolve) => {
    const worker = new Worker(__filename, {
      workerData: { 
        grid, 
        guardPos,
        trodden 
      }
    });
    
    worker.on('message', resolve);
    worker.on('error', (err) => {
      console.error(err);
      resolve(0);
    });
  });
}

function solve(grid) {
  const guard = getGuard(grid);
  
  // Using a hash map for path tracking
  const path = {};
  path[`${guard.pos[0]}:${guard.pos[1]}`] = [DIR_UP];
  
  // Faster iteration with direct boundary checks
  while (isInGrid(grid, guard.pos)) {
    if (!tick(grid, guard, path)) return null;
  }
  
  return path;
}

function tick(grid, guard, path) {
  move(grid, guard);
  
  // Faster key generation and lookup
  const pathKey = `${guard.pos[0]}:${guard.pos[1]}`;
  const directions = path[pathKey];
  
  // Check if cycle detected
  if (directions && directions.includes(guard.dir)) {
    return false;
  }
  
  if (isInGrid(grid, guard.pos)) {
    // Initialize array if needed (avoiding conditional assignment)
    if (!directions) {
      path[pathKey] = [];
    }
    
    path[pathKey]?.push(guard.dir);
  }
  
  return true;
}

function move(grid, guard) {
  const nextPosMod = nextPosMods[guard.dir];
  const nextPos = [guard.pos[0] + nextPosMod[0], guard.pos[1] + nextPosMod[1]];
  
  // Check if hitting wall
  if (isInGrid(grid, nextPos) && grid[nextPos[0]][nextPos[1]] === WALL) {
    guard.dir = turns[guard.dir];
  } else {
    guard.pos = nextPos;
  }
}

function getGuard(grid) {
  // Early return for optimization
  for (let i = 0; i < grid.length; i++) {
    const row = grid[i];
    for (let j = 0; j < row.length; j++) {
      if (row[j] === DIR_UP) {
        return { pos: [i, j], dir: DIR_UP };
      }
    }
  }
  return null;
}

// Inlined boundary check for performance
function isInGrid(grid, pos) {
  return (
    pos[0] >= 0 &&
    pos[0] < grid.length &&
    pos[1] >= 0 &&
    pos[1] < grid[0].length
  );
}

function getGrid() {
  // Read file once and parse efficiently
  const content = fs.readFileSync(filepath, { encoding: "utf-8" });
  
  // Split content into grid with minimal operations
  return content
    .trim()
    .split("\n")
    .map(line => line.split(""));
} 

File: 2024/day06Optimized/day06.py
import time
import multiprocessing as mp
from functools import lru_cache

start = time.time()

# Constants for better performance
DIR_UP = '^'
DIR_DOWN = 'v'
DIR_LEFT = '<'
DIR_RIGHT = '>'
WALL = '#'
OPEN = '.'

# Pre-computed lookup tables
next_pos_mods = {
    DIR_UP: (-1, 0),  # Using tuples is faster than lists
    DIR_DOWN: (1, 0),
    DIR_LEFT: (0, -1),
    DIR_RIGHT: (0, 1)
}

turns = {
    DIR_UP: DIR_RIGHT,
    DIR_DOWN: DIR_LEFT,
    DIR_LEFT: DIR_UP,
    DIR_RIGHT: DIR_DOWN
}

filename = '../inputs/day06.txt'

# Cache the grid parsing for reuse
_grid_cache = None

def get_grid():
    global _grid_cache
    if _grid_cache is None:
        with open(filename, 'r') as infile:
            _grid_cache = [list(line.strip()) for line in infile.readlines()]
    # Return a deep copy to avoid modifying the cached grid
    return [row[:] for row in _grid_cache]

# Fast inline grid boundary check
def is_in_grid(grid, pos):
    return 0 <= pos[0] < len(grid) and 0 <= pos[1] < len(grid[0])

# Optimized guard finder
def get_guard(grid):
    for i, row in enumerate(grid):
        try:
            # Using index is faster than iterating when we expect to find the item early
            j = row.index(DIR_UP)
            return {'pos': [i, j], 'dir': DIR_UP}
        except ValueError:
            continue
    return None

def move(grid, guard):
    next_pos_mod = next_pos_mods[guard['dir']]
    next_pos = [guard['pos'][0] + next_pos_mod[0], guard['pos'][1] + next_pos_mod[1]]
    
    if is_in_grid(grid, next_pos) and grid[next_pos[0]][next_pos[1]] == WALL:
        guard['dir'] = turns[guard['dir']]
    else:
        guard['pos'] = next_pos

def tick(grid, guard, path):
    move(grid, guard)
    path_key = tuple(guard['pos'])  # Tuples are hashable and faster than converting to strings
    
    # Fast path lookup and cycle detection
    if path_key in path and guard['dir'] in path[path_key]:
        return False
        
    if is_in_grid(grid, guard['pos']):
        if path_key not in path:
            path[path_key] = []
        path[path_key].append(guard['dir'])
    return True

def solve(grid):
    guard = get_guard(grid)
    if not guard:
        return None
        
    # Pre-allocate path dictionary with expected size
    path = {}
    path[tuple(guard['pos'])] = [DIR_UP]
    
    # Main solver loop
    while is_in_grid(grid, guard['pos']):
        if not tick(grid, guard, path):
            return None
    return path

def run_pt_1():
    grid = get_grid()
    return solve(grid)

# Worker function for multiprocessing
def process_coord(coord_data):
    pos, grid, guard_pos = coord_data
    
    # Skip guard starting position
    if pos[0] == guard_pos[0] and pos[1] == guard_pos[1]:
        return 0
        
    # Make a local copy of the grid
    local_grid = [row[:] for row in grid]
    
    # Modify the grid and check the solution
    local_grid[pos[0]][pos[1]] = WALL
    completed = solve(local_grid)
    
    return 0 if completed else 1

# Optimized part 2 with multiprocessing
def run_pt_2(trodden):
    grid = get_grid()
    guard = get_guard(grid)
    guard_pos = guard['pos']
    
    # Calculate number of processes to use (75% of CPU cores)
    num_cores = int(mp.cpu_count() * 0.75)
    if num_cores < 1:
        num_cores = 1
    
    # Prepare data for workers
    coords_data = [(pos, grid, guard_pos) for pos in trodden]
    
    # Use multiprocessing to process coordinates in parallel
    with mp.Pool(processes=num_cores) as pool:
        results = pool.map(process_coord, coords_data)
    
    # Sum the results
    count = sum(results)
    print(f'Part 2: {count}')
    return count

if __name__ == '__main__':
    # Part 1
    path = run_pt_1()
    trodden = list(path.keys())
    print(f"Part 1: {len(trodden)}")
    
    # Part 2
    run_pt_2(trodden)
    
    end = time.time()
    print(f'time: {end - start:.3f}s') 